### ISSUE TITLE
Intermittent problems with Kryo serialisation for crawls resumed from checkpoints

### ISSUE THREAD

**Issue by anjackson** [internetarchive/heritrix3#247]
I'm hitting problems when re-using crawl state (checkpoints). I get a lot of errors like:

```
WARNING: com.google.common.cache.LocalCache processPendingNotifications Exception thrown by removal listener [Tue Mar 19 12:07:00 GMT 2019]
java.lang.IllegalArgumentException: Can not set org.archive.modules.fetcher.FetchStats field org.archive.crawler.frontier.WorkQueue.substats to java.lang.Byte
        at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
        at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
        at sun.reflect.UnsafeObjectFieldAccessorImpl.set(UnsafeObjectFieldAccessorImpl.java:81)
        at java.lang.reflect.Field.set(Field.java:764)
        at com.esotericsoftware.kryo.serialize.FieldSerializer$CachedField.set(FieldSerializer.java:290)
        at com.esotericsoftware.kryo.serialize.FieldSerializer.readObjectData(FieldSerializer.java:209)
        at com.esotericsoftware.kryo.serialize.FieldSerializer.readObjectData(FieldSerializer.java:178)
        at com.esotericsoftware.kryo.Kryo.readObjectData(Kryo.java:512)
        at com.esotericsoftware.kryo.ObjectBuffer.readObjectData(ObjectBuffer.java:212)
        at org.archive.bdb.KryoBinding.entryToObject(KryoBinding.java:84)
        at com.sleepycat.collections.DataView.makeValue(DataView.java:595)
        at com.sleepycat.collections.DataCursor.getCurrentValue(DataCursor.java:349)
        at com.sleepycat.collections.DataCursor.initForPut(DataCursor.java:813)
        at com.sleepycat.collections.DataCursor.put(DataCursor.java:751)
        at com.sleepycat.collections.StoredContainer.putKeyValue(StoredContainer.java:321)
        at com.sleepycat.collections.StoredMap.put(StoredMap.java:279)
        at org.archive.util.ObjectIdentityBdbManualCache$1.onRemoval(ObjectIdentityBdbManualCache.java:119)
        at com.google.common.cache.LocalCache.processPendingNotifications(LocalCache.java:1954)
        at com.google.common.cache.LocalCache$Segment.runUnlockedCleanup(LocalCache.java:3457)
        at com.google.common.cache.LocalCache$Segment.postWriteCleanup(LocalCache.java:3433)
        at com.google.common.cache.LocalCache$Segment.put(LocalCache.java:2888)
        at com.google.common.cache.LocalCache.put(LocalCache.java:4146)
        at org.archive.util.ObjectIdentityBdbManualCache.dirtyKey(ObjectIdentityBdbManualCache.java:374)
        at org.archive.crawler.frontier.WorkQueue.makeDirty(WorkQueue.java:688)
        at org.archive.crawler.frontier.WorkQueueFrontier.processFinish(WorkQueueFrontier.java:1016)
        at org.archive.crawler.frontier.AbstractFrontier.finished(AbstractFrontier.java:569)
        at org.archive.crawler.framework.ToeThread.run(ToeThread.java:187)
```

One possible cause is that the Kryo serialisers are not getting set up right.

As I understand it, the reflection-based auto-registration magic attempts to register the classes needed, and as I understand [the documentation](https://github.com/EsotericSoftware/kryo/wiki/(outdated)-Kryo-v1#registered-ids) this saves storage space but relies on classes getting registered in a consistent order (so the same classes get the same IDs).

However, this registration appears to happen on the Spring `Lifecycle.start()` event, e.g. `org.archive.modules.net.BdbServerCache.start()` or `org.archive.crawler.frontier.WorkQueueFrontier.start()` and AFAICT nothing is explicitly enforcing the order of these events.

It looks like the latter leads to 

https://github.com/internetarchive/heritrix3/blob/05811705ed996122bea1f4e034c1ed5f7a07240f/modules/src/main/java/org/archive/modules/CrawlURI.java#L1808-L1811

(i.e. there we see `Byte` getting registered) and the former leads to 

https://github.com/internetarchive/heritrix3/blob/05811705ed996122bea1f4e034c1ed5f7a07240f/modules/src/main/java/org/archive/modules/net/CrawlServer.java#L319-L321

(i.e. there's `FetchStats`) which seems suspicious. However, in both cases, the autoregistered class is the second class to get registered, not the first, so it's not clear why this would be the case.

I'm having trouble understanding exactly what goes on with [Kryo 1 and thread context](https://github.com/EsotericSoftware/kryo/blob/1.05/src/com/esotericsoftware/kryo/Kryo.java#L377) and therefore whether the reference IDs are global or ThreadLocal or AutoKyro-instance scoped.

I'm left to assume I must have missed something, otherwise this would never have worked reliably at all!

**Comment by anjackson** [comment]
Hm, well it seems each store gets it's own `ObjectIdentityBdbManualCache` and each of those has it's own `AutoKryo` instance so this seems reasonably safe. i.e. the code should be able to reload the checkpoint if the code that registers the classes for that part of the system has not been changed?

**Comment by anjackson** [comment]
Note that Kryo operates a `ThreadLocal` cache that could perhaps have a bug in it? Long shot but we're running a very old Kryo (1.01 - not even the 1.0.5 I linked to above).

**Comment by anjackson** [comment]
FWIW, this appears to be Kryo @ 1.01 https://github.com/EsotericSoftware/kryo/blob/00902f65f222ebcd5bea197535f41cda1009b967/src/com/esotericsoftware/kryo/Kryo.java

**Comment by anjackson** [comment]
Hm okay Kryo registration appears to be scoped per instance and deterministic. No idea why this is throwing errors.

I’m going to run a full crawl but without resuming from a checkpoint and see if the issue turns up then.

**Comment by anjackson** [comment]
Other observations:

- Appears to happen only when starting from a checkpoint (possibly re-starting from a previously-used checkpoint?)
- No problems for a few hours, then suddenly the same error over and over a few times every second. The removal listener is only run once per item in the removal notification queue, so each error corresponds to a different entry. The cache in question has a maximum size of 10,000 and expiry time of five minutes.  
- The total number of queues in Heritrix was around 21,000 at when the problem first arose.
- The problem arose shortly after a significant increase in the number of `exhausted` queues most of which appeared to have been `inactive` state just prior to becoming `exhausted`. (Based on [this](https://github.com/internetarchive/heritrix3/wiki/Glossary#queue-states) that means going from 'out of rotation' to 'empty' pretty much immediately).
- Most of the exceptions are from the `processFinish`  trace (see above), but some are from the `schedule` trace (see below). If this issue arises from a runtime race-condition it seems likely to be between these two.
- The 'processFinished` trace is part of the end of the whole crawl process. The `schedule` trace occurs earlier in the `ToeThread`'s progression (during the disposition chain). The former operates on the `ToeThread`'s own `WorkQueue` whereas the latter is often operating on other queues.

```
java.lang.IllegalArgumentException: Can not set org.archive.modules.fetcher.FetchStats field org.archive.crawler.frontier.WorkQueue.substats to java.lang.Byte
        at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
        at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
        at sun.reflect.UnsafeObjectFieldAccessorImpl.set(UnsafeObjectFieldAccessorImpl.java:81)
        at java.lang.reflect.Field.set(Field.java:764)
        at com.esotericsoftware.kryo.serialize.FieldSerializer$CachedField.set(FieldSerializer.java:290)
        at com.esotericsoftware.kryo.serialize.FieldSerializer.readObjectData(FieldSerializer.java:209)
        at com.esotericsoftware.kryo.serialize.FieldSerializer.readObjectData(FieldSerializer.java:178)
        at com.esotericsoftware.kryo.Kryo.readObjectData(Kryo.java:512)
        at com.esotericsoftware.kryo.ObjectBuffer.readObjectData(ObjectBuffer.java:212)
        at org.archive.bdb.KryoBinding.entryToObject(KryoBinding.java:84)
        at com.sleepycat.collections.DataView.makeValue(DataView.java:595)
        at com.sleepycat.collections.DataCursor.getCurrentValue(DataCursor.java:349)
        at com.sleepycat.collections.DataCursor.initForPut(DataCursor.java:813)
        at com.sleepycat.collections.DataCursor.put(DataCursor.java:751)
        at com.sleepycat.collections.StoredContainer.putKeyValue(StoredContainer.java:321)
        at com.sleepycat.collections.StoredMap.put(StoredMap.java:279)
        at org.archive.util.ObjectIdentityBdbManualCache$1.onRemoval(ObjectIdentityBdbManualCache.java:119)
        at com.google.common.cache.LocalCache.processPendingNotifications(LocalCache.java:1954)
        at com.google.common.cache.LocalCache$Segment.runUnlockedCleanup(LocalCache.java:3457)
        at com.google.common.cache.LocalCache$Segment.postWriteCleanup(LocalCache.java:3433)
        at com.google.common.cache.LocalCache$Segment.put(LocalCache.java:2888)
        at com.google.common.cache.LocalCache.put(LocalCache.java:4146)
        at org.archive.util.ObjectIdentityBdbManualCache.dirtyKey(ObjectIdentityBdbManualCache.java:374)
        at org.archive.crawler.frontier.WorkQueue.makeDirty(WorkQueue.java:688)
        at org.archive.crawler.frontier.WorkQueueFrontier.sendToQueue(WorkQueueFrontier.java:409)
        at org.archive.crawler.frontier.WorkQueueFrontier.processScheduleAlways(WorkQueueFrontier.java:332)
        at org.archive.crawler.frontier.AbstractFrontier.receive(AbstractFrontier.java:549)
        at org.archive.crawler.util.SetBasedUriUniqFilter.add(SetBasedUriUniqFilter.java:85)
        at org.archive.crawler.frontier.WorkQueueFrontier.processScheduleIfUnique(WorkQueueFrontier.java:377)
        at org.archive.crawler.frontier.WorkQueueFrontier.schedule(WorkQueueFrontier.java:355)
        at org.archive.crawler.postprocessor.CandidatesProcessor.runCandidateChain(CandidatesProcessor.java:189)
        at org.archive.crawler.postprocessor.CandidatesProcessor.innerProcess(CandidatesProcessor.java:230)
        at org.archive.modules.Processor.innerProcessResult(Processor.java:175)
        at org.archive.modules.Processor.process(Processor.java:142)
        at org.archive.modules.ProcessorChain.process(ProcessorChain.java:131)
        at org.archive.crawler.framework.ToeThread.run(ToeThread.java:152)
```

**Comment by anjackson** [comment]
So, to my frustration, I think I'll just have to leave this.  To summarise:

- The `BdbWorkQueue` instances appear to have been loaded from disk once, but when the time comes to re-write them, the exact-same Kryo code appears to behave differently the second time it runs. 
- The very specific nature of the repeating error is difficult to understand. The sudden but specific mismatch of serialisation code seems to imply the Kryo registration map or related caches are getting damaged somehow.
- It is very difficult to see how corruption of the BDB data store itself would lead to such a 'neat' error.
- It is perhaps notable that `FetchStats` does not have a default constructor, and so serialisation relies on the _magic_ done in `AutoKryo` to invoke the constructor via reflection. This code appears to rely on `ReflectionFactory` being thread-safe, but there's no indication it should not be.
- Note also that the internal Kryo class registration map was [a HashMap in version 1.01](https://github.com/EsotericSoftware/kryo/blob/00902f65f222ebcd5bea197535f41cda1009b967/src/com/esotericsoftware/kryo/Kryo.java#L51) but [a ConcurrentHashMap by version 1.05](https://github.com/EsotericSoftware/kryo/blob/07b16e6872a04ed4102439139f39a1de14f80d09/src/com/esotericsoftware/kryo/Kryo.java#L52-L53). The [underlying issue](https://github.com/EsotericSoftware/kryo/issues/26) indicates that 1.01 registration is not thread-safe, but there does not appear to be any multi-threaded registration going on AFAICT (each thread has separate registration on startup I think?!). Even if there is, it's not clear how the de-serialisation could appear to work once and then fail the second time.

I will perhaps return to this if/when we look at updating BDB and Kryo to something other than deeply-ancient versions.

**Comment by anjackson** [comment]
Just observed in an (admittedly troubled by #245 issues) crawl:

```
WARNING: com.google.common.cache.LocalCache processPendingNotifications Exception thrown by removal listener (in thread 'ToeThread #135: https://www.theedinburghreporter.co.uk/category/news/holyrood/'; in processor 'candidates') [Wed Mar 27 14:12:15 GMT 2019] 
com.esotericsoftware.kryo.SerializationException: Unable to deserialize object of type: org.archive.crawler.frontier.BdbWorkQueue
	at com.esotericsoftware.kryo.Kryo.readObjectData(Kryo.java:514)
	at com.esotericsoftware.kryo.ObjectBuffer.readObjectData(ObjectBuffer.java:212)
	at org.archive.bdb.KryoBinding.entryToObject(KryoBinding.java:84)
	at com.sleepycat.collections.DataView.makeValue(DataView.java:595)
	at com.sleepycat.collections.DataCursor.getCurrentValue(DataCursor.java:349)
	at com.sleepycat.collections.DataCursor.initForPut(DataCursor.java:813)
	at com.sleepycat.collections.DataCursor.put(DataCursor.java:751)
	at com.sleepycat.collections.StoredContainer.putKeyValue(StoredContainer.java:321)
	at com.sleepycat.collections.StoredMap.put(StoredMap.java:279)
	at org.archive.util.ObjectIdentityBdbManualCache$1.onRemoval(ObjectIdentityBdbManualCache.java:119)
	at com.google.common.cache.LocalCache.processPendingNotifications(LocalCache.java:1954)
	at com.google.common.cache.LocalCache$Segment.runUnlockedCleanup(LocalCache.java:3457)
	at com.google.common.cache.LocalCache$Segment.postWriteCleanup(LocalCache.java:3433)
	at com.google.common.cache.LocalCache$Segment.put(LocalCache.java:2888)
	at com.google.common.cache.LocalCache.put(LocalCache.java:4146)
	at org.archive.util.ObjectIdentityBdbManualCache.dirtyKey(ObjectIdentityBdbManualCache.java:374)
	at org.archive.crawler.frontier.WorkQueue.makeDirty(WorkQueue.java:688)
	at org.archive.crawler.frontier.WorkQueueFrontier.sendToQueue(WorkQueueFrontier.java:409)
	at org.archive.crawler.frontier.WorkQueueFrontier.processScheduleAlways(WorkQueueFrontier.java:332)
	at org.archive.crawler.frontier.AbstractFrontier.receive(AbstractFrontier.java:549)
	at org.archive.crawler.util.SetBasedUriUniqFilter.add(SetBasedUriUniqFilter.java:85)
	at org.archive.crawler.frontier.WorkQueueFrontier.processScheduleIfUnique(WorkQueueFrontier.java:377)
	at org.archive.crawler.frontier.WorkQueueFrontier.schedule(WorkQueueFrontier.java:355)
	at org.archive.crawler.postprocessor.CandidatesProcessor.runCandidateChain(CandidatesProcessor.java:189)
	at org.archive.crawler.postprocessor.CandidatesProcessor.innerProcess(CandidatesProcessor.java:230)
	at org.archive.modules.Processor.innerProcessResult(Processor.java:175)
	at org.archive.modules.Processor.process(Processor.java:142)
	at org.archive.modules.ProcessorChain.process(ProcessorChain.java:131)
	at org.archive.crawler.framework.ToeThread.run(ToeThread.java:152)
Caused by: com.esotericsoftware.kryo.SerializationException: Encountered unregistered class ID: 23
	at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:381)
	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:469)
	at com.esotericsoftware.kryo.serialize.MapSerializer.readObjectData(MapSerializer.java:129)
	at com.esotericsoftware.kryo.serialize.FieldSerializer.readObjectData(FieldSerializer.java:198)
	at com.esotericsoftware.kryo.serialize.FieldSerializer.readObjectData(FieldSerializer.java:178)
	at com.esotericsoftware.kryo.serialize.FieldSerializer.readObjectData(FieldSerializer.java:198)
	at com.esotericsoftware.kryo.serialize.FieldSerializer.readObjectData(FieldSerializer.java:178)
	at com.esotericsoftware.kryo.Kryo.readObjectData(Kryo.java:512)
	... 28 more
```

**Comment by anjackson** [comment]
I believe this issue was to do with problems in managing checkpoints, and should have been resolved by #281.
